### PR TITLE
4.x shells

### PR DESCRIPTION
**Pull request by saeideng**

### CODE REVIEW

File: tests/TestCase/Command/BakeMigrationDiffCommandTest.php [hunk 1]
```diff
@@ -24,7 +26,22 @@
 use Migrations\Migrations;
 use Migrations\Test\TestCase\Shell\TestCompletionStringOutput;
 use Migrations\Test\TestCase\TestCase;
+use Migrations\Command\BakeMigrationDiffCommand;
 
+class CustomBakeMigrationDiffCommand extends BakeMigrationDiffCommand
+{
+    public $pathFragment = 'config/MigrationsDiff/';
```
Comment (saeideng) [inline tests/TestCase/Command/BakeMigrationDiffCommandTest.php]: I extended `BakeMigrationDiffCommand`  to set `pathFragment` and hacking  `getDumpSchema`

File: tests/bootstrap.php [hunk 2]
```diff
@@ -38,7 +38,7 @@
     define('DS', DIRECTORY_SEPARATOR);
 }
 define('CORE_PATH', $root . DS . 'vendor' . DS . 'cakephp' . DS . 'cakephp' . DS);
-define('ROOT', $root . DS . 'tests' . DS . 'test_app' . DS);
+define('ROOT', $root . DS . 'tests' . DS . 'test_app');
 define('APP_DIR', 'App');
 define('APP', ROOT . 'App' . DS);
```
Comment (dereuromark) [inline tests/bootstrap.php]: DS missing if you remove it above.

File: tests/test_app/App/BakeApplication.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace TestApp;
+
+use Cake\Console\CommandCollection;
+use Cake\Http\BaseApplication;
+use Cake\Http\MiddlewareQueue;
+use Cake\Routing\RouteBuilder;
+use Migrations\Test\Command\CustomBakeMigrationDiffCommand;
+
+class BakeApplication extends BaseApplication
```
Comment (dereuromark) [inline tests/test_app/App/BakeApplication.php]: Application should suffice here IMO.

File: tests/TestCase/Command/BakeMigrationDiffCommandTest.php [hunk 4]
```diff
@@ -53,6 +70,7 @@ public function setUp(): void
         $this->generatedFiles = [];
         $this->cleanupDatabase();
         $this->useCommandRunner();
+        $this->configApplication(Configure::read('App.namespace') . '\BakeApplication',[]);
```
Comment (dereuromark) [inline tests/TestCase/Command/BakeMigrationDiffCommandTest.php]: Then this one is not necessary :)

File: tests/TestCase/Command/BakeMigrationDiffCommandTest.php [hunk 5]
```diff
@@ -36,9 +36,34 @@ protected function getDumpSchema(Arguments $args)
...
         $diffConfigFolder = Plugin::path('Migrations') . 'tests' . DS . 'comparisons' . DS . 'Diff' . DS;
         $diffDumpPath = $diffConfigFolder . 'schema-dump-test_comparisons_' . env('DB') . '.lock';
-        $file = unserialize(file_get_contents($diffDumpPath));
 
-        return $file;
+        return unserialize(file_get_contents($diffDumpPath));
+    }
+
+}
+class CustomSimpleBakeMigrationDiffCommand extends BakeMigrationDiffCommand
```
Comment (dereuromark) [inline tests/TestCase/Command/BakeMigrationDiffCommandTest.php]: We shouldnt put those here, but inside TestApp dir etc.

File: tests/bootstrap.php (hunk 2 above)
Comment (saeideng) [reply tests/bootstrap.php]: I not found any usage :) but added

File: tests/test_app/App/BakeApplication.php (hunk 3 above)
Comment (saeideng) [reply tests/test_app/App/BakeApplication.php]: `extends Application` ??

File: tests/TestCase/Command/BakeMigrationDiffCommandTest.php (hunk 4 above)
Comment (saeideng) [reply tests/TestCase/Command/BakeMigrationDiffCommandTest.php]: needed for adding/run new commands

File: tests/test_app/App/BakeApplication.php (hunk 3 above)
Comment (dereuromark) [reply tests/test_app/App/BakeApplication.php]: class Application extends BaseApplication
see my plugins for example.

File: tests/TestCase/Command/BakeMigrationDiffCommandTest.php (hunk 5 above)
Comment (saeideng) [reply tests/TestCase/Command/BakeMigrationDiffCommandTest.php]: yes , but only uses in this testCase , tracking it when is not here could be hard later

File: tests/test_app/App/BakeApplication.php (hunk 3 above)
Comment (saeideng) [reply tests/test_app/App/BakeApplication.php]: there is a class as `Application` 
i can not use same name

File: tests/TestCase/Command/BakeMigrationDiffCommandTest.php (hunk 5 above)
Comment (dereuromark) [reply tests/TestCase/Command/BakeMigrationDiffCommandTest.php]: This violates psr loading rules and a bunch of other things.
As such still usually better to follow this strictly. One class per file.

Comment (saeideng) [reply tests/TestCase/Command/BakeMigrationDiffCommandTest.php]: and I no like to use custom class , but I not found better way

Comment (saeideng) [reply tests/TestCase/Command/BakeMigrationDiffCommandTest.php]: if phpstan is not happy with this , I  can put it to TestApp dir

File: tests/test_app/App/BakeApplication.php (hunk 3 above)
Comment (dereuromark) [reply tests/test_app/App/BakeApplication.php]: You are in TestApp
There are multiple ones in there already?

File: tests/TestCase/Command/BakeMigrationDiffCommandTest.php (hunk 5 above)
Comment (dereuromark) [reply tests/TestCase/Command/BakeMigrationDiffCommandTest.php]: classes in TestApp are free :)

File: tests/test_app/App/BakeApplication.php (hunk 3 above)
Comment (saeideng) [reply tests/test_app/App/BakeApplication.php]: yes

File: tests/TestCase/Command/BakeMigrationDiffCommandTest.php (hunk 5 above)
Comment (saeideng) [reply tests/TestCase/Command/BakeMigrationDiffCommandTest.php]: moved

File: tests/test_app/App/BakeApplication.php (hunk 3 above)
Comment (saeideng) [reply tests/test_app/App/BakeApplication.php]: merged with `Application`

File: src/Command/BakeMigrationDiffCommand.php [hunk 6]
```diff
@@ -451,28 +457,31 @@ protected function checkSync()
...
+        $newArgs[] = $name;
+
+        if (!empty($args->getOption('connection'))) {
+            $newArgs[] = '-c';
+            $newArgs[] = $args->getOption('connection');
         }
-        if (!empty($this->params['plugin'])) {
-            $dispatchCommand .= ' -p ' . $this->params['plugin'];
+
+        if (!empty($args->getOption('plugin'))) {
```
Comment (dereuromark) [inline src/Command/BakeMigrationDiffCommand.php:472]: !empty() should not be used around methods, so just without it here.

**Comment by saeideng** [comment]
running individual  tests is OK but whole testCase no
i think there is conflict between testCases 
database is clear on this testCase (`migrationTest`) but again ....

**Comment by saeideng** [comment]
ready to review and merge

tests failures are unrelated to subject of this PR
and we could ignore those for now

File: tests/TestCase/MigrationsTest.php [hunk 7]
```diff
@@ -75,6 +81,12 @@ public function setUp(): void
         $adapter->setConnection($connection);
 
         $tables = (new Collection($this->Connection))->listTables();
+        foreach ($tables as $table) {
+            if ($table === 'phinxlog') {
```
Comment (dereuromark) [inline tests/TestCase/MigrationsTest.php:85]: I would use a simliar regex than I did for bake. For this this is OK though.

**Pull request merged**